### PR TITLE
More cwm objects

### PR DESCRIPTION
**Pull request by jreidinger**

### CODE REVIEW

File: src/lib/y2network/widgets/boot_protocol.rb [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+      def valid_netmask(ip, mask)
+        valid_mask = false
+        mask = mask[1..-1] if mask.start_with?("/")
+
+        if Yast::IP.Check4(ip) && (Yast::Netmask.Check4(mask) || Yast::Netmask.CheckPrefix4(mask))
+          valid_mask = true
+        elsif Yast::IP.Check6(ip) && Yast::Netmask.Check6(mask)
+          valid_mask = true
+        else
+          log.warn "IP address #{ip} is not valid"
```
Comment (mchf) [inline src/lib/y2network/widgets/boot_protocol.rb]: a bit strange when method is named ```valid_netmask``` ;-) May be it should be ```valid_static_ip_setup``` or something like that

File: src/lib/y2network/widgets/boot_protocol.rb [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+        ipa = Yast::UI.QueryWidget(:bootproto_ipaddr, :Value)
+        if !ipa.empty? && !Yast::IP.Check(ipa)
+          Yast::Popup.Error(_("No valid IP address."))
+          Yast::UI.SetFocus(:bootproto_ipaddr)
+          return false
+        end
+
+        mask = Yast::UI.QueryWidget(:bootproto_netmask, :Value)
+        if ipa != "" && mask != "" && !valid_netmask(ipa, mask)
```
Comment (mchf) [inline src/lib/y2network/widgets/boot_protocol.rb]: NP: these checks for empty values can be removed. Validating functions should catch that ... or should be fixed to catch that.

File: src/lib/y2network/widgets/boot_protocol.rb [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        else
+          raise "Unexpected boot protocol value #{Yast::UI.QueryWidget(:bootproto, :Value).inspect}"
+        end
+      end
+
+      def validate
+        return true if value != :bootproto_static
+
+        ipa = Yast::UI.QueryWidget(:bootproto_ipaddr, :Value)
+        if !ipa.empty? && !Yast::IP.Check(ipa)
```
Comment (mchf) [inline src/lib/y2network/widgets/boot_protocol.rb]: ```Yast::IP.Check("")``` -> false => ```ipa.empty?``` can be removed

File: src/lib/y2network/widgets/boot_protocol.rb [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+                  [
+                    Item(Id(:bootproto_dhcp_both), _("DHCP both version 4 and 6")),
+                    Item(Id(:bootproto_dhcp_v4), _("DHCP version 4 only")),
+                    Item(Id(:bootproto_dhcp_v6), _("DHCP version 6 only"))
+                  ]
+                )
+              )
+            ),
+            VBox(
+              # TODO : Stat ... Assigned
```
Comment (mchf) [inline src/lib/y2network/widgets/boot_protocol.rb]: ?

Comment (jreidinger) [reply src/lib/y2network/widgets/boot_protocol.rb]: I have no idea what it means, it was in original code.

File: src/lib/y2network/widgets/boot_protocol.rb (hunk 3 above)
Comment (jreidinger) [reply src/lib/y2network/widgets/boot_protocol.rb]: good tip.

File: src/lib/y2network/widgets/boot_protocol.rb [hunk 5]
```diff
@@ -0,0 +1,364 @@
+require "cwm/custom_widget"
```
Comment (mchf) [inline src/lib/y2network/widgets/boot_protocol.rb]: NP: I'd expect ```require "yast"``` too

File: src/lib/y2network/widgets/boot_protocol.rb (hunk 2 above)
Comment (jreidinger) [reply src/lib/y2network/widgets/boot_protocol.rb]: that is a bit problematic, as it means that we cannot leave tab without specifying it.

There are two possible approaches:

1. do not allow to leave tab until everything is correct
2. builder that validates its content and write error popup when user try to finish creating such device

Not such which one is better as both have advantages and disadvantages.

File: src/lib/y2network/widgets/boot_protocol.rb (hunk 5 above)
Comment (jreidinger) [reply src/lib/y2network/widgets/boot_protocol.rb]: yes, true

File: src/lib/y2network/widgets/boot_protocol.rb (hunk 1 above)
Comment (jreidinger) [reply src/lib/y2network/widgets/boot_protocol.rb]: well, original name is validPrefixOrNetmask(ipa, mask) so I just try to shorten it :)

File: src/lib/y2network/widgets/boot_protocol.rb (hunk 2 above)
Comment (jreidinger) [reply src/lib/y2network/widgets/boot_protocol.rb]: ah, now I see. `ipa` does not need to be checked. and mask is first time checked here and here is validating function and really catched.

File: src/lib/y2network/widgets/boot_protocol.rb (hunk 4 above)
Comment (mchf) [reply src/lib/y2network/widgets/boot_protocol.rb]: that was the point .... doesn't make sense to keep comments which cannot be understand ;-)

**Pull request merged**